include/cocaine/rpc/slot/generic.hpp

```diff
@@ -0,0 +1,59 @@
+/*
+    Copyright (c) 2011-2014 Andrey Sibiryov <me@kobology.ru>
+    Copyright (c) 2011-2014 Other contributors as noted in the AUTHORS file.
+
+    This file is part of Cocaine.
+
+    Cocaine is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef COCAINE_IO_GENERIC_SLOT_HPP
+#define COCAINE_IO_GENERIC_SLOT_HPP
+
+#include "cocaine/rpc/slot.hpp"
+
+#include "cocaine/tuple.hpp"
+
+#include <asio/system_error.hpp>
+
+namespace cocaine { namespace io {
+
+template<class Event, class F>
+struct generic_slot:
```

Comment:
In fact I need this for locator refactoring, but I've tested it on logging service in plugins.
Example:
```    
typedef io::basic_slot<io::base_log::get> get_slot;
on<io::base_log::get>([&](const std::vector<hpack::header_t>&,
                                            get_slot::tuple_type&& args,
                                            get_slot::upstream_type&& upstream) -> get_slot::call_result_type {
        auto metafilter = impl->get_metafilter(std::get<0>(args));
        std::shared_ptr<const get_slot::dispatch_type> dispatch = std::make_shared<const named_logging_t>(
                impl->logger, std::move(std::get<0>(args)), std::move(metafilter));
        return boost::make_optional(std::move(dispatch));
});
```

instead of

```
class logging_v2_t::logging_slot_t : public io::basic_slot<io::base_log::get> {
public:
    typedef io::base_log::get event_type;
    typedef typename io::basic_slot<event_type>::dispatch_type dispatch_type;
    typedef typename io::basic_slot<event_type>::tuple_type tuple_type;
    // This one is type of upstream.
    typedef typename io::basic_slot<event_type>::upstream_type upstream_type;

    logging_slot_t(logging_v2_t& _parent) : parent(_parent) {}

    virtual boost::optional<std::shared_ptr<const dispatch_type>> operator()(const std::vector<hpack::header_t>&,
                                                                             tuple_type&& args,
                                                                             upstream_type&&) {
        return tuple::invoke(std::forward<tuple_type>(args), [&](std::string name) {
            auto metafilter = parent.impl->get_metafilter(name);
            std::shared_ptr<const dispatch_type> dispatch = std::make_shared<const named_logging_t>(
                parent.impl->logger, std::move(name), std::move(metafilter));
            return boost::make_optional(std::move(dispatch));
        });
    }

private:
    logging_v2_t& parent;
};
```